src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20240103_143700.java

```diff
@@ -0,0 +1,497 @@
+            log.info("Will migrate {} student programming exercise participations in batch.", studentParticipationPage.getNumberOfElements());
+            executorService.submit(() -> {
+                migrateStudents(studentParticipationPage.toList());
+                studentCounter.addAndGet(studentParticipationPage.getNumberOfElements());
+                logProgress(studentCounter.get(), studentCount, threadCount, 1, "student");
+            });
+        }
+
+        log.info("Submitted all student participations to thread pool for migration.");
+
+        shutdown(executorService, TIMEOUT_IN_HOURS, ERROR_MESSAGE);
+        log.info("Finished migrating programming exercises and student participations");
+        evaluateErrorList(programmingExerciseRepository);
+    }
+
+    private void logProgress(long doneCount, long totalCount, long threadCount, long reposPerEntry, String migrationType) {
+        final double percentage = ((double) doneCount / totalCount) * 100;
+        log.info("Migrated {}/{} {} participations ({}%)", doneCount, totalCount, migrationType, String.format("%.2f", percentage));
+        log.info("Estimated time remaining: {} for {} repositories", TimeLogUtil.formatDuration(getRestDurationInSeconds(doneCount, totalCount, reposPerEntry, threadCount)),
+                migrationType);
+    }
+
+    private long getRestDurationInSeconds(final long done, final long total, final long reposPerEntry, final long threads) {
+        final long stillTodo = total - done;
+        final long timePerEntry = ESTIMATED_TIME_PER_REPOSITORY * reposPerEntry;
+        return (stillTodo * timePerEntry) / threads;
+    }
+
+    private String migrateTestRepo(ProgrammingExercise programmingExercise) throws URISyntaxException {
+        return cloneRepositoryFromBitbucketAndMoveToLocalVCS(programmingExercise, programmingExercise.getTestRepositoryUri(), programmingExercise.getBranch());
+    }
+
+    /**
+     * Migrate auxiliary repositories of the given programming exercise.
+     *
+     * @param programmingExercise the programming exercise to migrate the auxiliary repositories for
+     * @param oldBranch           the old branch of the programming exercise
+     */
+    private void migrateAuxiliaryRepositories(ProgrammingExercise programmingExercise, String oldBranch) {
+        for (var repo : getAuxiliaryRepositories(programmingExercise.getId())) {
+            try {
+                if (repo.getRepositoryUri() == null) {
+                    log.error("Repository URI is null for auxiliary repository with id {}, cant migrate", repo.getId());
+                    continue;
+                }
+                var url = cloneRepositoryFromBitbucketAndMoveToLocalVCS(programmingExercise, repo.getRepositoryUri(), oldBranch);
+                if (url == null) {
+                    log.error("Failed to migrate auxiliary repository for programming exercise with id {}, keeping the url in the database", programmingExercise.getId());
```

Comment:
How is the url kept in the database here? Also should these also get added to the error list?

Reply:
we don't override the url, that only happens in the other branch, so that IF any errors occur we can check manually. I added the solutionParticipation to the errorlist now, thanks